LMDB "Transaction ingester aborted" problem in 1.22.0

After running some time, the node seems to stop ingesting.
Await tx causes exception.

The node in question is running on Linux in AWS ECS.
Starting a new instance will restore from checkpoint and indexing works again. 
Don't know the exact pattern that causes the initial problem.


Log has entries with the following exception:

```
java.lang.Exception: Transaction ingester aborted.
at xtdb.node$await_tx.invokeStatic(node.clj:60)
at xtdb.node$await_tx.invoke(node.clj:36)
at xtdb.node.XtdbNode.await_tx(node.clj:158)
at xtdb.node.XtdbNode.await_tx(node.clj:155)
... application code...
Caused by: java.lang.AssertionError: Assert failed: LMDB write lock timeout
(pos? stamp)
at xtdb.lmdb$acquire_write_lock.invokeStatic(lmdb.clj:39)
at xtdb.lmdb$acquire_write_lock.invokePrim(lmdb.clj)
at xtdb.lmdb.LMDBKv.store(lmdb.clj:238)
at xtdb.fork.ForkedKvIndexStoreTx.commit_index_tx(fork.clj:188)
at xtdb.tx.InFlightTx.commit(tx.clj:356)
at xtdb.tx$__GT_tx_ingester$fn__45494$txs_index_fn__45502.invoke(tx.clj:567)
at clojure.lang.AFn.applyToHelper(AFn.java:154)
at clojure.lang.AFn.applyTo(AFn.java:144)
at clojure.core$apply.invokeStatic(core.clj:665)
at clojure.core$apply.invoke(core.clj:660)
at xtdb.tx$__GT_tx_ingester$fn__45494$submit_job_BANG___45495$fn__45496.invoke(tx.clj:544)
at clojure.lang.AFn.call(AFn.java:18)
... 4 common frames omitted
```


The earliest related message seems to be:
```
level | ERROR
logger | xtdb.tx
message | Ingester error occurred
thread | xtdb-tx-processor-1

java.lang.AssertionError: Assert failed: LMDB write lock timeout
(pos? stamp)
at xtdb.lmdb$acquire_write_lock.invokeStatic(lmdb.clj:39)
at xtdb.lmdb$acquire_write_lock.invokePrim(lmdb.clj)
at xtdb.lmdb.LMDBKv.store(lmdb.clj:238)
at xtdb.fork.ForkedKvIndexStoreTx.commit_index_tx(fork.clj:188)
at xtdb.tx.InFlightTx.commit(tx.clj:356)
at xtdb.tx$__GT_tx_ingester$fn__45494$txs_index_fn__45502.invoke(tx.clj:567)
at clojure.lang.AFn.applyToHelper(AFn.java:154)
at clojure.lang.AFn.applyTo(AFn.java:144)
at clojure.core$apply.invokeStatic(core.clj:665)
at clojure.core$apply.invoke(core.clj:660)
at xtdb.tx$__GT_tx_ingester$fn__45494$submit_job_BANG___45495$fn__45496.invoke(tx.clj:544)
at clojure.lang.AFn.call(AFn.java:18)
at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
at java.base/java.lang.Thread.run(Thread.java:833)
```

After that all the await tx calls start failing.

--- a/modules/lmdb/project.clj
+++ b/modules/lmdb/project.clj
@@ -11,6 +11,13 @@
 
   :scm {:dir "../.."}
 
+  :java-source-paths ["src"]
+  :javac-options ["-source" "8" "-target" "8"
+                  "-XDignore.symbol.file"
+                  "-Xlint:all,-options,-path"
+                  "-Werror"
+                  "-proc:none"]
+
   :dependencies [[org.clojure/clojure]
                  [org.clojure/tools.logging]
                  [com.xtdb/xtdb-core]

--- a/modules/lmdb/project.clj
+++ b/modules/lmdb/project.clj
@@ -1,23 +1,6 @@
 (defproject com.xtdb/xtdb-lmdb "<inherited>"
   :description "XTDB LMDB"
 
-  :plugins [[lein-parent "0.3.8"]]
-
-  :parent-project {:path "../../project.clj"
-                   :inherit [:version :repositories :deploy-repositories
-                             :managed-dependencies
-                             :pedantic? :global-vars
-                             :license :url :pom-addition]}
-
-  :scm {:dir "../.."}
-
-  :java-source-paths ["src"]
-  :javac-options ["-source" "8" "-target" "8"
-                  "-XDignore.symbol.file"
-                  "-Xlint:all,-options,-path"
-                  "-Werror"
-                  "-proc:none"]
-
   :dependencies [[org.clojure/clojure]
                  [org.clojure/tools.logging]
                  [com.xtdb/xtdb-core]
@@ -28,4 +11,36 @@
                  [org.lwjgl/lwjgl-lmdb "3.3.1" :classifier "natives-macos" :native-prefix ""]
                  [org.lwjgl/lwjgl "3.3.1" :classifier "natives-macos-arm64" :native-prefix ""]
                  [org.lwjgl/lwjgl-lmdb "3.3.1" :classifier "natives-macos-arm64" :native-prefix ""]
-                 [org.lwjgl/lwjgl-lmdb "3.3.1"]])
+                 [org.lwjgl/lwjgl-lmdb "3.3.1"]]
+
+  :plugins [[lein-javadoc "0.3.0"]
+            [lein-parent "0.3.8"]]
+
+  :parent-project {:path "../../project.clj"
+                   :inherit [:version :repositories :deploy-repositories
+                             :managed-dependencies
+                             :pedantic? :global-vars
+                             :license :url :pom-addition]}
+
+  :scm {:dir "../.."}
+
+  :java-source-paths ["src"]
+
+  :javac-options ["-source" "8" "-target" "8"
+                  "-XDignore.symbol.file"
+                  "-Xlint:all,-options,-path"
+                  "-Werror"
+                  "-proc:none"]
+
+  :javadoc-opts {:package-names ["xtdb"]
+                 :output-dir "target/javadoc/out"
+                 :additional-args ["-windowtitle" "XTDB LMDB Javadoc"
+                                   "-quiet"
+                                   "-Xdoclint:none"
+                                   "-link" "https://docs.oracle.com/javase/8/docs/api/"
+                                   "-link" "https://www.javadoc.io/static/org.clojure/clojure/1.10.3"]}
+
+  :classifiers {:sources {:prep-tasks ^:replace []}
+                :javadoc {:prep-tasks ^:replace ["javadoc"]
+                          :omit-source true
+                          :filespecs ^:replace [{:type :path, :path "target/javadoc/out"}]}})

--- a/modules/lmdb/src/xtdb/MapResizeSync.java
+++ b/modules/lmdb/src/xtdb/MapResizeSync.java
@@ -23,9 +23,9 @@ public class MapResizeSync extends AbstractQueuedSynchronizer {
     // Notes:
     // I wanted to avoid the .java, but couldn't easily use (proxy) in clj because protected methods cannot be called.
 
-    // Reason for queued synchronizer to get fairness policy and platform blocking for LMDB locks without implementing our own spin.
+    // Reason for queued synchronizer to get fairness policy and platform blocking for LMDB read locks without implementing our own spin.
     // originally the stamped lock that was in use did not enforce fairness / write priority which could cause ingestion to wait too long
-    // to get lock.
+    // to get the lock.
     public MapResizeSync() {
         super();
     }
@@ -33,28 +33,23 @@ public class MapResizeSync extends AbstractQueuedSynchronizer {
     @Override
     protected boolean tryAcquire(int signal) {
         // note: not re-entrant, not yet necessary
-        assert signal == -1;
-        if (compareAndSetState(0, -1)) {
-            setExclusiveOwnerThread(Thread.currentThread());
-            return true;
+        // instead of queuing, writes always spin
+        // they will then probabilistically barge ahead of writes, if they go on the queue it will cause deadlocks
+        // when reads nest on another thread.
+        for(;;) {
+            if (compareAndSetState(0, -1)) {
+                return true;
+            }
         }
-        return false;
     }
 
     @Override
     protected boolean tryRelease(int signal) {
         assert signal == -1;
-        if (!isHeldExclusively()) {
-            throw new IllegalMonitorStateException("Map lock released on-non owner thread");
+        if (compareAndSetState(-1, 0)) {
+            return true;
         }
-        setState(0);
-        setExclusiveOwnerThread(null);
-        return true;
-    }
-
-    @Override
-    protected boolean isHeldExclusively() {
-        return getExclusiveOwnerThread() == Thread.currentThread();
+        throw new IllegalMonitorStateException("Map lock not acquired");
     }
     @Override
     protected int tryAcquireShared(int signal) {

--- a/modules/lmdb/src/xtdb/MapResizeSync.java
+++ b/modules/lmdb/src/xtdb/MapResizeSync.java
@@ -1,0 +1,85 @@
+package xtdb;
+
+import java.util.concurrent.locks.AbstractQueuedSynchronizer;
+
+/**
+ * Provides synchronization for LMDB map resizes, see lmdb.clj (new-mapsize-sync).
+ * This class is an implementation detail of the lmdb module and may change without notice.
+ *
+ *
+ * Magic numbers:
+ *
+ * Lock state (atomic int):
+ * -1 write lock held for resize
+ * 0 no open transactions, no resize lock (default)
+ * 1+ number of open transactions
+ *
+ * Signal arg to acquire/release methods is 1 for 'open transaction' or -1 for 'map resize lock'.
+ */
+public class MapResizeSync extends AbstractQueuedSynchronizer {
+
+    private static final long serialVersionUID = -1672790154983593262L;
+
+    // Notes:
+    // I wanted to avoid the .java, but couldn't easily use (proxy) in clj because protected methods cannot be called.
+
+    // Reason for queued synchronizer to get fairness policy and platform blocking for LMDB locks without implementing our own spin.
+    // originally the stamped lock that was in use did not enforce fairness / write priority which could cause ingestion to wait too long
+    // to get lock.
+    public MapResizeSync() {
+        super();
+    }
+
+    @Override
+    protected boolean tryAcquire(int signal) {
+        // note: not re-entrant, not yet necessary
+        assert signal == -1;
+        if (compareAndSetState(0, -1)) {
+            setExclusiveOwnerThread(Thread.currentThread());
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    protected boolean tryRelease(int signal) {
+        assert signal == -1;
+        if (!isHeldExclusively()) {
+            throw new IllegalMonitorStateException("Map lock released on-non owner thread");
+        }
+        setState(0);
+        setExclusiveOwnerThread(null);
+        return true;
+    }
+
+    @Override
+    protected boolean isHeldExclusively() {
+        return getExclusiveOwnerThread() == Thread.currentThread();
+    }
+    @Override
+    protected int tryAcquireShared(int signal) {
+        assert signal == 1;
+        if (hasQueuedPredecessors()) {
+            return -1;
+        }
+        int st = getState();
+        if (st >= 0 && compareAndSetState(st, st+1)) {
+            return st+1;
+        }
+        return -1;
+    }
+
+    @Override
+    protected boolean tryReleaseShared(int signal) {
+        assert signal == 1;
+        for(;;) {
+            int st = getState();
+            if (st < 1) {
+                throw new IllegalMonitorStateException("Release of map resize read lock did not correspond to an acquire.");
+            }
+            if (compareAndSetState(st, st-1)) {
+                return true;
+            }
+        }
+    }
+}

--- a/modules/lmdb/src/xtdb/lmdb.clj
+++ b/modules/lmdb/src/xtdb/lmdb.clj
@@ -305,8 +305,7 @@
         env (env-create)
         mapsize-lock (StampedLock.)]
     (try
-      (when env-set-maxreaders
-        (env-set-maxreaders env env-maxreaders))
+      (env-set-maxreaders env env-maxreaders)
       (env-open env db-dir env-flags)
       (when env-mapsize
         (env-set-mapsize env env-mapsize))

--- a/modules/lmdb/src/xtdb/lmdb.clj
+++ b/modules/lmdb/src/xtdb/lmdb.clj
@@ -44,7 +44,9 @@
   When a resize is needed the (kv/store) call will acquire an exclusive (resize) lock, ensuring no concurrent transactions,
   new transactions must wait (block) for the resize to finish. See calls of (acquire-map-resize-permit sync), (release-map-resize-permit sync)
 
-  The reason over StampedLock (the original control mechanism) is to provide a fairness/barging policy so that writers are prioritised over readers, and that older readers get a chance to run if under high contention.
+  The reason over StampedLock (the original control mechanism) is to provide a fairness/barging policy for readers if under high contention, writes are probabilistically
+  preferred to readers, but there is no fairness between writers. This approach avoids deadlocks when reads nest while a resize acquire is spinning.
+
   We cannot use a ReentrantReadWriteLock as open transactions lifetimes are not tied to any particular thread.
 
   NOTE: resize permits must be released on acquiring thread, transaction permits can be released on any thread."

--- a/modules/lmdb/src/xtdb/lmdb.clj
+++ b/modules/lmdb/src/xtdb/lmdb.clj
@@ -14,13 +14,12 @@
            java.io.Closeable
            [java.nio.file Files Path]
            java.nio.file.attribute.FileAttribute
-           java.util.concurrent.locks.StampedLock
-           java.util.concurrent.TimeUnit
            org.agrona.concurrent.UnsafeBuffer
            org.agrona.ExpandableDirectByteBuffer
            [org.lwjgl.system MemoryStack MemoryUtil]
            [org.lwjgl.util.lmdb LMDB MDBEnvInfo MDBStat MDBVal]
-           (clojure.lang IDeref)))
+           (clojure.lang IDeref)
+           (xtdb MapResizeSync)))
 
 (set! *unchecked-math* :warn-on-boxed)
 
@@ -34,10 +33,29 @@
 (defn- env-set-mapsize [^long env ^long size]
   (success? (LMDB/mdb_env_set_mapsize env size)))
 
-(defn- acquire-write-lock ^long [^StampedLock mapsize-lock]
-  (let [stamp (.tryWriteLock mapsize-lock 120 TimeUnit/SECONDS)]
-    (assert (pos? stamp) "LMDB write lock timeout")
-    stamp))
+(defn new-mapsize-sync
+  "Provides concurrency control to LMDB access to avoid races/segfault when map resizes are required.
+
+  Occasionally writes to LMDB will encounter a 'MDB_MAP_FULL' error, xtdb responds by increasing the size of the lmdb map, retrying the write.
+  This can only be done if there are no open transactions, otherwise the JVM will segfault.
+
+  We track open transactions by acquiring/releasing 'open transaction permits' via (acquire-open-transaction-permit sync), (release-open-transaction-permit sync).
+
+  When a resize is needed the (kv/store) call will acquire an exclusive (resize) lock, ensuring no concurrent transactions,
+  new transactions must wait (block) for the resize to finish. See calls of (acquire-map-resize-permit sync), (release-map-resize-permit sync)
+
+  The reason over StampedLock (the original control mechanism) is to provide a fairness/barging policy so that writers are prioritised over readers, and that older readers get a chance to run if under high contention.
+  We cannot use a ReentrantReadWriteLock as open transactions lifetimes are not tied to any particular thread.
+
+  NOTE: resize permits must be released on acquiring thread, transaction permits can be released on any thread."
+  ^MapResizeSync []
+  (MapResizeSync.))
+
+;; see new-mapsize-sync def above
+(defn- acquire-map-resize-permit [^MapResizeSync mapsize-sync] (.acquireInterruptibly mapsize-sync -1))
+(defn- release-map-resize-permit [^MapResizeSync mapsize-sync] (.release mapsize-sync -1))
+(defn- acquire-open-transaction-permit [^MapResizeSync mapsize-sync] (.acquireSharedInterruptibly mapsize-sync 1))
+(defn- release-open-transaction-permit [^MapResizeSync mapsize-sync] (.releaseShared mapsize-sync 1))
 
 ;; TODO: Note, this has to be done when there are no open
 ;; transactions. Also, when file reached 4Gb it crashed. MDB_WRITEMAP
@@ -76,13 +94,13 @@
 (defn- new-transaction-no-lock ^Closeable [env flags]
   (closeable (txn-begin env flags) txn-close))
 
-(defn- new-transaction-owning-lock ^Closeable [^StampedLock mapsize-lock env flags]
-  (let [stamp (.readLock mapsize-lock)]
-    (try
-      (closeable (txn-begin env flags) #(try (txn-close %) (finally (.unlock mapsize-lock stamp))))
-      (catch Throwable t
-        (.unlock mapsize-lock stamp)
-        (throw t)))))
+(defn- new-transaction-owning-lock ^Closeable [mapsize-sync env flags]
+  (acquire-open-transaction-permit mapsize-sync)
+  (try
+    (closeable (txn-begin env flags) #(try (txn-close %) (finally (release-open-transaction-permit mapsize-sync))))
+    (catch Throwable t
+      (release-open-transaction-permit mapsize-sync)
+      (throw t))))
 
 (defn- env-create []
   (with-open [stack (MemoryStack/stackPush)]
@@ -112,9 +130,9 @@
     (.mkdirs file)
     (success? (LMDB/mdb_env_copy env (.getAbsolutePath file)))))
 
-(defn- dbi-open [mapsize-lock ^long env]
+(defn- dbi-open [mapsize-sync ^long env]
   (with-open [stack (MemoryStack/stackPush)
-              tx (new-transaction-owning-lock mapsize-lock env LMDB/MDB_RDONLY)]
+              tx (new-transaction-owning-lock mapsize-sync env LMDB/MDB_RDONLY)]
     (let [^long txn @tx
           ip (.mallocInt stack 1)
           ^CharSequence name nil]
@@ -135,9 +153,12 @@
       (success? rc)
       (UnsafeBuffer. (.mv_data kv) 0 (.mv_size kv)))))
 
-(defn- cursor-put [env dbi kvs]
+(defn- cursor-put [env dbi mapsize-sync kvs]
   (with-open [stack (MemoryStack/stackPush)
-              tx (new-transaction-no-lock env 0)
+              ;; if we are writing under an exclusive lock during a map resize, mapsize-sync is nil.
+              tx (if mapsize-sync
+                   (new-transaction-owning-lock mapsize-sync env 0)
+                   (new-transaction-no-lock env 0))
               cursor (closeable (cursor-open dbi @tx) cursor-close)]
     (let [kv (MDBVal/malloc stack)
           dv (MDBVal/malloc stack)
@@ -214,9 +235,9 @@
 (def ^:dynamic ^{:tag 'long} *mapsize-increase-factor* 1)
 (def ^:const max-mapsize-increase-factor 32)
 
-(defn- put-if-map-not-full [env dbi kvs]
+(defn- put-if-map-not-full [env dbi mapsize-sync kvs]
   (try
-    (cursor-put env dbi kvs)
+    (cursor-put env dbi mapsize-sync kvs)
     true
     (catch ExceptionInfo e
       (if (= LMDB/MDB_MAP_FULL (:error (ex-data e)))
@@ -224,7 +245,8 @@
         (throw e)))))
 
 (defn- put-resizing-if-map-full [env dbi kvs]
-  (when-not (put-if-map-not-full env dbi kvs)
+  ;; assume we have exclusive access if resizing, nil signals to the (put-if-map-not-full) to use no lock.
+  (when-not (put-if-map-not-full env dbi nil kvs)
     (binding [*mapsize-increase-factor* (* 2 *mapsize-increase-factor*)]
       (when (> *mapsize-increase-factor* max-mapsize-increase-factor)
         (throw (IllegalStateException. "Too large size of key values to store at once.")))
@@ -232,14 +254,19 @@
       ;; non-tail recur to carry binding for max-mapsize-increase error
       (put-resizing-if-map-full env dbi kvs))))
 
-(defrecord LMDBKv [db-dir env env-flags dbi ^StampedLock mapsize-lock sync? cp-job]
+(defrecord LMDBKv [db-dir env env-flags dbi mapsize-sync sync? cp-job]
   kv/KvStore
   (store [this kvs]
-    (with-open [_ (closeable (acquire-write-lock mapsize-lock) #(.unlock mapsize-lock %))]
-      (put-resizing-if-map-full env dbi kvs)))
+    (when-not (put-if-map-not-full env dbi mapsize-sync kvs)
+      ;; on MDB_MAP_FULL we must ensure no open transactions to resize, or segfault - see (new-mapsize-sync)
+      (acquire-map-resize-permit mapsize-sync)
+      (try
+        (put-resizing-if-map-full env dbi kvs)
+        (finally
+          (release-map-resize-permit mapsize-sync)))))
 
   (new-snapshot [_]
-    (->LMDBKvSnapshot env dbi (new-transaction-owning-lock mapsize-lock env LMDB/MDB_RDONLY)))
+    (->LMDBKvSnapshot env dbi (new-transaction-owning-lock mapsize-sync env LMDB/MDB_RDONLY)))
 
   (compact [_])
 
@@ -249,7 +276,7 @@
 
   (count-keys [_]
     (with-open [stack (MemoryStack/stackPush)
-                tx (new-transaction-owning-lock mapsize-lock env LMDB/MDB_RDONLY)]
+                tx (new-transaction-owning-lock mapsize-sync env LMDB/MDB_RDONLY)]
       (let [stat (MDBStat/malloc stack)]
         (LMDB/mdb_stat @tx dbi stat)
         (.ms_entries stat))))
@@ -268,11 +295,11 @@
 
   Closeable
   (close [_]
-    (let [stamp (acquire-write-lock mapsize-lock)]
-      (try
-        (env-close env)
-        (finally
-          (.unlock mapsize-lock stamp))))
+    (acquire-map-resize-permit mapsize-sync)
+    (try
+      (env-close env)
+      (finally
+        (release-map-resize-permit mapsize-sync)))
     (xio/try-close cp-job)))
 
 (def ^:private cp-format
@@ -303,7 +330,7 @@
                                 0
                                 no-sync-env-flags)))
         env (env-create)
-        mapsize-lock (StampedLock.)]
+        mapsize-sync (new-mapsize-sync)]
     (try
       (env-set-maxreaders env env-maxreaders)
       (env-open env db-dir env-flags)
@@ -312,11 +339,11 @@
       (let [kv-store (map->LMDBKv {:db-dir db-dir
                                    :env env
                                    :env-flags env-flags
-                                   :dbi (dbi-open mapsize-lock env)
-                                   :mapsize-lock mapsize-lock
+                                   :dbi (dbi-open mapsize-sync env)
+                                   :mapsize-sync mapsize-sync
                                    :sync? sync?})]
         (cond-> kv-store
-          checkpointer (assoc :cp-job (cp/start checkpointer kv-store {::cp/cp-format cp-format}))))
+                checkpointer (assoc :cp-job (cp/start checkpointer kv-store {::cp/cp-format cp-format}))))
       (catch Throwable t
         (env-close env)
         (throw t)))))

